# DYNAMIC PROGRAMMING

RULES :
1. Build Solution
2. Optimise it to Recursion solution
3. Let's see if it has OPTIMAL SUBSTRUCTURE ?
4. DP if YES else NO

In [1]:
# One best Decorator of all time FAV ❤️
from functools import lru_cache

# Example 1 
## Fibanocci Series

```
    fib series -> 0, 1, 1, 2, 3, 5, 8,...
             n -> 0, 1, 2, 3, 4, 5, 6,...

                   fib(n-1) + fib(n-2)     ; if n > 1 
function fib =  {  1                       ; if n = 1     
                   0                       ; if n = 0 
```

### Step 1 :
Build Solution 

In [2]:
def fib(n):
    if n == 0:
        return 0
    elif n == 1:
        return 1
    else:
        f0 = 0
        f1 = 1
        for i in range(n-1):
            temp = f1 + f0
            f0 = f1
            f1 = temp
        return f1
fib(6)

8

### Step 2 :
Make Solution Recursive

In [3]:
fibcalls = 0
def fib(n):
    global fibcalls
    fibcalls += 1
    if n == 0:
        return 0
    elif n == 1:
        return 1
    return fib(n-1) + fib(n-2)
fib(6),fibcalls

(8, 25)

### Step 3 :
Check for Overlapping Subproblems 
```

                    fib(4)
                    /      \
              fib(3)        fib(2)  
              /     \         /  \ 
          fib(2)   fib(1) fib(1) fib(0)      
          /    \
      fib(1)  fib(0)      

```

In [4]:
fibcalls = 0
fib(4),fibcalls

(3, 9)

Out of 9 Calls there is unique calls are just 5 

Therefore there is Optimal Substructure

### Step 4 :
## DP Methods :
1. Top Down   - recursive memoization
2. Bottum UP  - itterative Tabulation 
3. Cacheeeeeeeeeeeeeeeeeeeeeeeeeeeeee


## DP : Method 1
Top Down - Recursive - Memoization

In [5]:
fib_DP_topDowncalls = 0
DP_TD = {0:0,1:1}
def fib_DP_topDown(n):
    global fib_DP_topDowncalls
    global DP_TD
    fib_DP_topDowncalls += 1
    if n in DP_TD:
        return DP_TD[n]
    DP_TD[n] = (fib_DP_topDown(n-1) if n-1 not in (0,1) else DP_TD[n-1]) + (fib_DP_topDown(n-2) if n-2 not in (0,1) else DP_TD[n-2] )
    return DP_TD[n]
# fib_DP_topDown(6),fib_DP_topDowncalls

In [6]:
fibcalls = 0
fib(4)
fib_DP_topDowncalls = 0
fib_DP_topDown(4)
fibcalls,fib_DP_topDowncalls 

(9, 4)

## DP : Method 2
Bottom Up - Itterative - Tabultion

In [7]:
fib_DP_BottomUpcalls = 0
DP_BU = {0:0,1:1}
def fib_DP_bottomUp(n):
    global fib_DP_BottomUpcalls
    global DP_BU
    fib_DP_BottomUpcalls += 1
    for i in range(2,n+1):
        if i not in DP_BU:
            DP_BU[i] = DP_BU[i-1] + DP_BU[i-2]
    return DP_BU[n]
# fib_DP_bottomUp(6),fib_DP_BottomUpcalls

In [8]:
fibcalls = 0
fib(4)
fib_DP_topDowncalls = 0
fib_DP_topDown(4)
fib_DP_bottomUp(4)
fibcalls,fib_DP_topDowncalls,fib_DP_BottomUpcalls

(9, 1, 1)

### Conclusion 
AS we can observe there is no recursive calls for Bottom Up so prefer BU than TP coming to Python Exclusive since there stack will give exceptions in TD for large input

## DP : Method 3
Cacheeeeeeeeeeeeeeeeeeeeeeee

Recursive - Simple FAST

In [9]:
from functools import lru_cache

fibcalls_cache = 0

@lru_cache
def fib_cache(n):
    global fibcalls_cache
    fibcalls_cache += 1
    if n == 0:
        return 0
    elif n == 1:
        return 1
    return (fib_DP_topDown(n-1) if n-1 not in (0,1) else n-1) + (fib_DP_topDown(n-2) if n-2 not in (0,1) else n-2 )

# fib_cache(6),fibcalls_cache

## Test On all versions

In [10]:
K = 50

In [11]:
fibcalls = 0
fib(K)

12586269025

In [16]:
fib_DP_topDowncalls = 0
fib_DP_topDown(K)

12586269025

In [17]:
fib_DP_BottomUpcalls = 0
fib_DP_bottomUp(K)

12586269025

In [18]:
fibcalls_cache = 0
fib_cache(K)

12586269025

In [15]:
fibcalls,fib_DP_topDowncalls,fib_DP_BottomUpcalls,fibcalls_cache

(40730022147, 95, 1, 1)

### Result of fib :

    ALGO                   CALLS            TIME         SPACE  
    fibcalls               40730022147      132m 56s     2n
    fib_DP_topDowncalls    95               0.1s         2n
    fib_DP_BottomUpcalls   1                1.6s         n
    fibcalls_cache         1                0.9s         2n

    Runinng Again 
    fib_DP_topDowncalls    1                0.1s         
    fib_DP_BottomUpcalls   1                0.1s         
    fibcalls_cache         1                0.4s         
